src/metabase/sync_database/cached_values.clj

```diff
@@ -0,0 +1,136 @@
+        distinct-values (when (field-values-below-low-cardinality-threshold? non-nil-values)
+                          non-nil-values)]
+    (if (seq distinct-values)
+      (assoc field-stats :values distinct-values)
+      field-stats)))
+
+(defn make-field-extractor
+  "Make a generic implementation of `analyze-table`."
+  {:style/indent 1}
+  [driver & {:keys [calculate-row-count?]
+             :or   {calculate-row-count?  true}}]
+  (fn [driver table new-field-ids]
+    {:fields    (for [id new-field-ids]
+                  (extract-field-values (field/Field id) {:id id}))}))
+
+(defn generic-analyze-table
+  "An implementation of `analyze-table` using the defaults (`default-field-avg-length` and `field-percent-urls`)."
+  [driver table new-field-ids]
+  ((make-field-extractor driver) driver table new-field-ids))
+
+;;;; End of driver stuff
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; call drivers, get field values, save them in application DB
+(defn cache-table-data-shape!
+  "Analyze the data shape for a single `Table`."
+  [driver {table-id :id, :as table}]
+  (let [new-field-ids (db/select-ids field/Field, :table_id table-id, :visibility_type [:not= "retired"], :last_analyzed nil)]
+    ;; TODO: this call should include the database
+    (when-let [table-stats (u/prog1 (driver/analyze-table driver table new-field-ids)
+                             (when <>
+                               (schema/validate i/AnalyzeTable <>)))]
+      ;; update table row count @sameer should this be saved here where it happens to be available or calculated as part of making
+      ;; the fingerprints in analyze
+      (when (:row_count table-stats)
+        (db/update! table/Table table-id, :rows (:row_count table-stats)))
+
+      ;; update individual fields
+      (doseq [{:keys [id preview-display values]} (:fields table-stats)]
+        ;; handle field values, setting them if applicable otherwise clearing them
+        (if (and id values (pos? (count (filter identity values))))
+          (field-values/save-field-values! id values)
+          (field-values/clear-field-values! id))))))
+
+(defn cache-field-values-for-table!
+  "Save the field values for each field in this database"
+  [table]
+  (cache-table-data-shape! (->> table
+                                table/database
+                                :id
+                                driver/database-id->driver)
+                           table))
+
+(defn cache-data-shape-for-tables!
+  "Perform in-depth analysis on the data shape for all `Tables` in a given DATABASE.
+   This is dependent on what each database driver supports, but includes things like cardinality testing and table row counting.
+   The bulk of the work is done by the `(analyze-table ...)` function on the IDriver protocol."
```

Comment:
Copy-pasta-ing this out-of-date docstring from the old sync code to here and to the `analyze-data-shape-for-tables!` function doesn't do much to make the new sync code easier to work with. 😓